### PR TITLE
satty: init at 0.7.0

### PR DESCRIPTION
**Pull request by pinpox**
## Description of changes

Added package for satty, a screenshot annotation tool inspired by Swappy and Flameshot. 
Init from latest main commit, as it contains a fix (thanks @a-kenji !) for the build that is not in the release and breaks the build.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/sa/satty/Cargo.lock
Comment (a-kenji) [inline pkgs/by-name/sa/satty/Cargo.lock:1]: Since you are using the fixed main branch, you can drop this lock file.

**Review by a-kenji — APPROVED** [review]
LGTM, after dropping the lock file.

File: pkgs/by-name/sa/satty/Cargo.lock
Comment (pinpox) [reply pkgs/by-name/sa/satty/Cargo.lock:1]: Fixed

File: pkgs/by-name/sa/satty/package.nix [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+  buildInputs = [
+    cairo
+    gdk-pixbuf
+    glib
+    gtk4
+    libadwaita
+    pango
+  ];
+
```
Comment (hmenke) [inline pkgs/by-name/sa/satty/package.nix:40]:
```suggestion

  passthru .updateScript = unstableGitUpdater { };

```

File: pkgs/by-name/sa/satty/package.nix [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  meta = with lib; {
+    description = "Satty - Modern Screenshot Annotation. A Screenshot Annotation Tool inspired by Swappy and Flameshot";
+    homepage = "https://github.com/gabm/Satty/tree/main";
```
Comment (hmenke) [inline pkgs/by-name/sa/satty/package.nix:43]:
```suggestion
    homepage = "https://github.com/gabm/Satty";
```

File: pkgs/by-name/sa/satty/package.nix [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+, gdk-pixbuf
+, glib
+, gtk4
+, libadwaita
+, pango
+}:
+
+rustPlatform.buildRustPackage rec {
+  pname = "satty";
+  version = "main-2023-11-21";
```
Comment (hmenke) [inline pkgs/by-name/sa/satty/package.nix:16]:
```suggestion
  version = "unstable-2023-11-21";
```

**Review by hmenke — CHANGES_REQUESTED** [review]

File: pkgs/by-name/sa/satty/package.nix [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    libadwaita
+    pango
+  ];
+
+  meta = with lib; {
+    description = "Satty - Modern Screenshot Annotation. A Screenshot Annotation Tool inspired by Swappy and Flameshot";
+    homepage = "https://github.com/gabm/Satty/tree/main";
+    license = licenses.mpl20;
+    maintainers = with maintainers; [ pinpox ];
+    mainProgram = "satty";
```
Comment (hmenke) [inline pkgs/by-name/sa/satty/package.nix:46]: Does it build on macOS? If not, please add
```suggestion
    mainProgram = "satty";
    platforms = lib.platforms.linux;
```

File: pkgs/by-name/sa/satty/package.nix (hunk 3 above)
Comment (hmenke) [reply pkgs/by-name/sa/satty/package.nix:16]: Why are you packaging the unstable version at all? There are tagged releases.

Comment (a-kenji) [reply pkgs/by-name/sa/satty/package.nix:16]: The tagged release would mean vendoring the lock file and carrying a patch. The fix for that is only after the tagged release.

Comment (hmenke) [reply pkgs/by-name/sa/satty/package.nix:16]: This should work just fine.
```nix
{
  version = "0.7.0";

  src = fetchFromGitHub {
    owner = "gabm";
    repo = "Satty";
    rev = "v${version}";
    hash = "sha256-x2ljheG7ZqaeiPersC/e8Er2jvk5TJs65Y3N1GjTiNU=";
  };

  cargoPatches = [
    (fetchpatch {
      name = "fix-Cargo.lock";
      url = "https://github.com/gabm/Satty/commit/39be6ddce264552df971e949a6a3175b102530b2.patch";
      hash = "sha256-GUHupZE1A7AmXvZ8WvRzBkQyH7qlMTetBjHuakfIZ7w=";
    })
  ];

  cargoHash = "sha256-0GsbWd/gpKZm7nNXkuJhB02YKUj3XCrSfpRA9KBXydU=";
}
```

Comment (a-kenji) [reply pkgs/by-name/sa/satty/package.nix:16]: Ah, thanks! Yes that does seem like it would work.

File: pkgs/by-name/sa/satty/package.nix [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  version = "unstable-2023-11-21";
+
+  src = fetchFromGitHub {
+    owner = "gabm";
+    repo = "Satty";
+    rev = "e1ad952e51c2f4a0f768d0d89418ae30eeda2cc3";
+    hash = "sha256-rpBOWZOwZcon8/09pAIqhBwU8uD6Nuh4gGjAzGFO0Y0=";
+  };
+
+  cargoHash = "sha256-UxFhs5I99kwo/+gY0k8Z20H8x1MJsG97SrswOqFJfAk=";
```
Comment (hmenke) [inline pkgs/by-name/sa/satty/package.nix:25]: As discussed, please package the release version together with the cargo patch.
```suggestion
  version = "0.7.0";

  src = fetchFromGitHub {
    owner = "gabm";
    repo = "Satty";
    rev = "v${version}";
    hash = "sha256-x2ljheG7ZqaeiPersC/e8Er2jvk5TJs65Y3N1GjTiNU=";
  };

  cargoPatches = [
    (fetchpatch {
      name = "fix-Cargo.lock";
      url = "https://github.com/gabm/Satty/commit/39be6ddce264552df971e949a6a3175b102530b2.patch";
      hash = "sha256-GUHupZE1A7AmXvZ8WvRzBkQyH7qlMTetBjHuakfIZ7w=";
    })
  ];

  cargoHash = "sha256-0GsbWd/gpKZm7nNXkuJhB02YKUj3XCrSfpRA9KBXydU=";
```

File: pkgs/by-name/sa/satty/package.nix [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  buildInputs = [
+    cairo
+    gdk-pixbuf
+    glib
+    gtk4
+    libadwaita
+    pango
+  ];
+
+  passthru .updateScript = unstableGitUpdater { };
```
Comment (hmenke) [inline pkgs/by-name/sa/satty/package.nix:41]: When switching to the released version, this can go again.
```suggestion
```

**Review by hmenke — CHANGES_REQUESTED** [review]

**Comment by pinpox** [comment]
@hmenke Hope I got it right after the back-and-forth. I've squashed the branch for clarity, it now packages release version + patch as suggested

**Comment by hmenke** [comment]
Result of `nixpkgs-review pr 268980` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>satty</li>
  </ul>
</details>

**Review by hmenke — APPROVED** [review]

File: pkgs/by-name/sa/satty/package.nix [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    cairo
+    gdk-pixbuf
+    glib
+    gtk4
+    libadwaita
+    pango
+  ];
+
+  meta = with lib; {
+    description = "Satty - A Screenshot Annotation Tool inspired by Swappy and Flameshot";
```
Comment (donovanglover) [inline pkgs/by-name/sa/satty/package.nix:52]:
```suggestion
    description = "A screenshot annotation tool inspired by Swappy and Flameshot";
```
See: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#meta-attributes

**Review by donovanglover — COMMENTED** [review]
Thanks for packaging this! I played around with it and everything seems to work. Would love to see the desktop file and icon added as well.

cc @ElysaSrc since she expressed interest in packaging this

**Comment by ElysaSrc** [comment]
Thanks for the notification @donovanglover ! 
Unfortunatly I didn't had time to write the derivation yet, but this looks great ! 
I'll give it a try tonight.

**Comment by hmenke** [comment]
Satty already ships with a desktop item which only needs to be installed by the appropriate hook. The icon needs to be copied manually.

```diff
--- a/satty.nix
+++ b/satty.nix
@@ -10,6 +10,7 @@
 , libadwaita
 , pango
 , fetchpatch
+, copyDesktopItems
 }:
 
 rustPlatform.buildRustPackage rec {
@@ -35,6 +36,7 @@ rustPlatform.buildRustPackage rec {
   cargoHash = "sha256-0GsbWd/gpKZm7nNXkuJhB02YKUj3XCrSfpRA9KBXydU=";
 
   nativeBuildInputs = [
+    copyDesktopItems
     pkg-config
     wrapGAppsHook4
   ];
@@ -48,6 +50,12 @@ rustPlatform.buildRustPackage rec {
     pango
   ];
 
+  postInstall = ''
+    install -Dt $out/share/icons/hicolor/scalable/apps/ assets/satty.svg
+  '';
+
+  desktopItems = [ "satty.desktop" ];
+
   meta = with lib; {
     description = "Satty - A Screenshot Annotation Tool inspired by Swappy and Flameshot";
     homepage = "https://github.com/gabm/Satty";
```

**Review by Janik-Haag — COMMENTED** [review]
seems good once you address the currently open comments

**Review by hmenke — APPROVED** [review]
LGTM!
Maybe change the commit message to `satty: init at 0.7.0`.

File: pkgs/by-name/sa/satty/package.nix (hunk 7 above)
Comment (pinpox) [reply pkgs/by-name/sa/satty/package.nix]: Applied and squashed

**Review by hmenke — APPROVED** [review]

**Comment by pinpox** [comment]
> LGTM! Maybe change the commit message to `satty: init at 0.7.0`.

I applied all the suggestions. Also renamed commit and PR to reflect the stable version.

**Comment by hmenke** [comment]
Result of `nixpkgs-review pr 268980` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>satty</li>
  </ul>
</details>

**Comment by Janik-Haag** [comment]
Result of `nixpkgs-review pr 268980` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>satty</li>
  </ul>
</details>

**Review by Janik-Haag — APPROVED** [review]

**Comment by gabm** [comment]
(*original author of satty here*)

@pinpox @a-kenji thank you for this contribution! Would you mind adding installation instruction to Satty's README for other fellow nix users?

**Pull request merged**